usermods/Battery/usermod_v2_Battery.h

```diff
@@ -0,0 +1,783 @@
+#pragma once
+
+#include "wled.h"
+#include "battery_defaults.h"
+
+/*
+ * Usermod by Maximilian Mewes
+ * Mail: mewes.maximilian@gmx.de
+ * GitHub: itCarl
+ * Date: 25.12.2022
+ * If you have any questions, please feel free to contact me.
+ */
+class UsermodBattery : public Usermod 
+{
+  private:
+    // battery pin can be defined in my_config.h
+    int8_t batteryPin = USERMOD_BATTERY_MEASUREMENT_PIN;
+    // how often to read the battery voltage
+    unsigned long readingInterval = USERMOD_BATTERY_MEASUREMENT_INTERVAL;
+    unsigned long nextReadTime = 0;
+    unsigned long lastReadTime = 0;
+    // battery min. voltage
+    float minBatteryVoltage = USERMOD_BATTERY_MIN_VOLTAGE;
+    // battery max. voltage
+    float maxBatteryVoltage = USERMOD_BATTERY_MAX_VOLTAGE;
+    // all battery cells summed up
+    unsigned int totalBatteryCapacity = USERMOD_BATTERY_TOTAL_CAPACITY;
+    // raw analog reading 
+    float rawValue = 0.0;
+    // calculated voltage            
+    float voltage = maxBatteryVoltage;
+    // mapped battery level based on voltage
+    int8_t batteryLevel = 100;
+    // offset or calibration value to fine tune the calculated voltage
+    float calibration = USERMOD_BATTERY_CALIBRATION;
+    
+    // time left estimation feature
+    // bool calculateTimeLeftEnabled = USERMOD_BATTERY_CALCULATE_TIME_LEFT_ENABLED;
+    // float estimatedTimeLeft = 0.0;
+
+    // auto shutdown/shutoff/master off feature
+    bool autoOffEnabled = USERMOD_BATTERY_AUTO_OFF_ENABLED;
+    int8_t autoOffThreshold = USERMOD_BATTERY_AUTO_OFF_THRESHOLD;
+
+    // low power indicator feature
+    bool lowPowerIndicatorEnabled = USERMOD_BATTERY_LOW_POWER_INDICATOR_ENABLED;
+    int8_t lowPowerIndicatorPreset = USERMOD_BATTERY_LOW_POWER_INDICATOR_PRESET;
+    int8_t lowPowerIndicatorThreshold = USERMOD_BATTERY_LOW_POWER_INDICATOR_THRESHOLD;
+    int8_t lowPowerIndicatorReactivationThreshold = lowPowerIndicatorThreshold+10;
+    int8_t lowPowerIndicatorDuration = USERMOD_BATTERY_LOW_POWER_INDICATOR_DURATION;
+    bool lowPowerIndicationDone = false;
+    unsigned long lowPowerActivationTime = 0; // used temporary during active time
+    int8_t lastPreset = 0;
+
+    bool initDone = false;
+    bool initializing = true;
+
+    // strings to reduce flash memory usage (used more than twice)
+    static const char _name[];
+    static const char _readInterval[];
+    static const char _enabled[];
+    static const char _threshold[];
+    static const char _preset[];
+    static const char _duration[];
+    static const char _init[];
+    
+
+    // custom map function
+    // https://forum.arduino.cc/t/floating-point-using-map-function/348113/2
+    double mapf(double x, double in_min, double in_max, double out_min, double out_max) 
+    {
+      return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+    }
+
+    float truncate(float val, int8_t dec) 
+    {
```

Comment:
How often will you run this function?
Because it seems like a very inefficient implementation. For example pow() is really really slow, and I don't understand why to use round and then again check delta == 5.

Ok if you only need this function a few times per second. Otherwise I can give you a more efficient code without pow(), so just ask me.


How often will you run this function?
Because it seems like a very inefficient implementation. For example pow() is really really slow, and I don't understand why to use round and then again check delta == 5.

Ok if you only need this function a few times per second. Otherwise I can give you a more efficient code without pow(), so just ask me.

Reply:
Good point. I just wondered myself. May have been an old relict from a test i did or something.
(fixed it, see next commit)